File: delightreading-server/src/main/java/com/delightreading/user/UserAccountEntity.java
package com.delightreading.user;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.PropertyNamingStrategy;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.Cascade;
import org.hibernate.annotations.Fetch;
import org.hibernate.annotations.FetchMode;

import javax.persistence.*;
import java.io.Serializable;
import java.time.Instant;
import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

import static javax.persistence.GenerationType.IDENTITY;

@Entity
@Table(name = "user_account")

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
public class UserAccountEntity implements Serializable {

    ////////// Base {{
    @Id
    @GeneratedValue(strategy = IDENTITY)
    Long sid;

    @Column(name = "uid")
    String uid;

    @Column(name = "status")
    String status;

    @Column(name = "created_by")
    String createdBy;

    @Column(name = "created_at")
    Instant createdAt;

    @Column(name = "updated_by")
    String updatedBy;

    @Column(name = "updated_at")
    Instant updatedAt;


    @PrePersist
    public void prePersist() {
        if (uid == null) {
            uid = UUID.randomUUID().toString();
        }
        createdAt = Instant.now();
        // createdBy = LoggedUser.get();
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = Instant.now();
        // updatedBy = LoggedUser.get();
    }
    ////////// }} Base

    @Column(name = "username")
    String username;

    @ElementCollection
    @CollectionTable(name = "user_account_emails", joinColumns = @JoinColumn(name = "account_uid", referencedColumnName = "uid"))
    //@Fetch(FetchMode.SUBSELECT)
    @Fetch(FetchMode.JOIN)
    @Cascade(org.hibernate.annotations.CascadeType.DELETE)
    @Column(name = "email")
    List<String> emails;

    @Column(name = "nickname")
    String nickname;

    @Column(name = "given_name")
    String givenName;

    @Column(name = "family_name")
    String familyName;

    @Column(name = "middle_name")
    String middleName;

    @Column(name = "date_of_birth")
    LocalDate dateOfBirth;

    @Column(name = "picture_uri")
    String pictureUri;

    @Column(name = "locale")
    String locale;

    @Column(name = "timezone")
    String timezone;

//    @OneToOne(mappedBy = "account", cascade = CascadeType.ALL,
//            fetch = FetchType.LAZY, optional = false)
//    @JoinColumn(name="profile_uid", referencedColumnName = "uid")
//    UserProfileEntity profile;

}

File: delightreading-backend/src/main/kotlin/com/delightreading/user/model/UserAccountEntity.kt
package com.delightreading.user.model

import org.hibernate.annotations.Cascade
import org.hibernate.annotations.Fetch
import org.hibernate.annotations.FetchMode
import java.io.Serializable
import java.time.Instant
import java.time.LocalDate
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "user_account")
data class UserAccountEntity(

    //<editor-fold desc="Common Entity Base">
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var sid: Long? = null,

    @Column(name = "uid")
    var uid: String? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    var status: UserAccountStatus = UserAccountStatus.ACTIVE,

    @Column(name = "created_by")
    var createdBy: String? = null,

    @Column(name = "created_at")
    var createdAt: Instant = Instant.now(),

    @Column(name = "updated_by")
    var updatedBy: String? = null,

    @Column(name = "updated_at")
    var updatedAt: Instant? = null,
    //</editor-fold>

    @Column(name = "username")
    var username: String? = null,

    @ElementCollection
    @CollectionTable(
        name = "user_account_emails",
        joinColumns = arrayOf(JoinColumn(name = "account_uid", referencedColumnName = "uid"))
    )
    //@Fetch(FetchMode.SUBSELECT)
    @Fetch(FetchMode.JOIN)
    @Cascade(org.hibernate.annotations.CascadeType.DELETE)
    @Column(name = "email")
    internal var emails: List<String>? = null,

    @Column(name = "nickname")
    internal var nickname: String? = null,

    @Column(name = "given_name")
    internal var givenName: String? = null,

    @Column(name = "family_name")
    internal var familyName: String? = null,

    @Column(name = "middle_name")
    internal var middleName: String? = null,

    @Column(name = "date_of_birth")
    internal var dateOfBirth: LocalDate? = null,

    @Column(name = "picture_uri")
    internal var pictureUri: String? = null,

    @Column(name = "locale")
    internal var locale: String? = null,

    @Column(name = "timezone")
    internal var timezone: String? = null

) : Serializable {

    //<editor-fold desc="Common Entity Base">
    @PrePersist
    fun prePersist() {
        if (uid == null) {
            uid = UUID.randomUUID().toString()
        }
        createdAt = Instant.now()
        // createdBy = LoggedUser.get();
    }

    @PreUpdate
    fun preUpdate() {
        updatedAt = Instant.now()
        // updatedBy = LoggedUser.get();
    }
    //</editor-fold>

}